### PR TITLE
[inductor] do not use `ceil` in `arange` ref

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #95773
* #95698

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -187,6 +187,12 @@ def _print_FloorDiv(self, expr):
         div = self.paren(self.doprint(div))
         return f"({x} / {div})"
 
+    def _print_ceiling(self, expr):
+        assert len(expr.args) == 1
+        return (
+            f"static_cast<int64_t>(std::ceil({self.paren(self._print(expr.args[0]))}))"
```
Comment (nkaretnikov) [inline torch/_inductor/codegen/cpp.py:193]: Used for array access, so we need to cast to integer type.

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -80,6 +80,12 @@ def _print_floor(self, expr):
         assert len(expr.args) == 1
         return f"tl.libdevice.floor({self.paren(self._print(expr.args[0]))})"
 
+    def _print_ceiling(self, expr):
+        assert len(expr.args) == 1
+        return (
+            f"tl.libdevice.ceil({self.paren(self._print(expr.args[0]))}).to(tl.int64)"
+        )
```
Comment (nkaretnikov) [inline torch/_inductor/codegen/triton.py:87]: Used with floordiv, which doesn't support non-int args in Triton (see `assert False` in `floordiv` in `triton/language/semantic.py`).

Example expr:
```
  tmp0 = 4*((x0 // (tl.libdevice.ceil(((-3.0/4.0) + ((1.0/4.0)*ks1))))))

```

**Comment by nkaretnikov** [comment]
Example codegen changes before (-) vs after (+):

CPP:
```
-                        out_ptr0[i2 + (i1*((ceiling(ks2/4 - 3/4))*(ceiling(ks2/4 - 3/4)))) + (16*i0*ks1*((ceiling(ks2/4 - 3/4))*(ceiling(ks2/4 - 3/4))))] = tmp6;
+                        out_ptr0[i2 + (i1*((static_cast<int64_t>(std::ceil(((-3.0/4.0) + ((1.0/4.0)*ks2)))))*(static_cast<int64_t>(std::ceil(((-3.0/4.0) + ((1.0/4.0)*ks2))))))) + (16*i0*ks1*((static_cast<int64_t>(std::ceil(((-3.0/4.0) + ((1.0/4.0)*ks2)))))*(static_cast<int64_t>(std::ceil(((-3.0/4.0) + ((1.0/4.0)*ks2)))))))] = tmp6;
```

Triton:
```
-    tmp0 = 4*((x0 // (math.ceil(((-3.0/4.0) + ((1.0/4.0)*ks1))))))
+    tmp0 = 4*((x0 // (tl.libdevice.ceil(((-3.0/4.0) + ((1.0/4.0)*ks1))).to(tl.int64))))
```

Old code would never work due to identifiers being incorrect.

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (ezyang) [inline torch/_inductor/codegen/cpp.py:193]: You don't need to parenthensize into function argument

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (ezyang) [reply torch/_inductor/codegen/triton.py:87]: This is oddly asymmetric with the floor above, which doesn't convert to int64. Why the int64 conversion?

Comment (nkaretnikov) [reply torch/_inductor/codegen/triton.py:87]: I think this is due to where these might show up based on our tests. This ceil expr is passed to floor. If floor in Triton handled non-int args, this wouldn't be necessary. Note that I also have to cast to int type on the C++ side, so at least that's consistent.

**Comment by ngimel** [comment]
I'd rather we fix iota decomp to not have ceil https://github.com/pytorch/pytorch/blob/d9f822b566bf2ae22010becb7277ccb381adba08/torch/_refs/__init__.py#L4369
```
length =(xend - xstart + xstep - 1) // xstep
```

**Comment by peterbell10** [comment]
Yeah for the int64 case using trunc-div is reasonable, though slightly more complex because `xstep` can be negative:
https://github.com/pytorch/pytorch/blob/d9f822b566bf2ae22010becb7277ccb381adba08/aten/src/ATen/native/RangeFactories.cpp#L183-L184

But note that you still need to handle `ceil` for all other dtypes since we use floating point to calculate the length.

**Comment by ngimel** [comment]
Non-integral types should never appear in shape computation, and even if they do, we should hoist it outside of triton kernel.  `(tl.libdevice.ceil(((-3.0/4.0) + ((1.0/4.0)*ks1))))` doesn't do what one thinks it is doing for `ks1` that is 2**26 or so.

**Comment by peterbell10** [comment]
> Non-integral types should never appear in shape computation

This output shape of `at::arange` uses `double` to calculate the output shape except for the special case of `int64`. To avoid off-by-one errors in shape, we need to do the same in the decomposition.

> `(tl.libdevice.ceil(((-3.0/4.0) + ((1.0/4.0)*ks1))))` doesn't do what one thinks it is doing for `ks1` that is `2**26` or so.

It should be calculated on the host anyway, so using `float64` shouldn't be an issue.

**Comment by ngimel** [comment]
It should be, but this PR puts this computation on the device.

**Comment by nkaretnikov** [comment]
`test_sizehint_issue1_dynamic_shapes`: before (-) vs after (+):

```
-                        auto tmp3 = static_cast<long>(4*(i2 % (ceiling(ks2/4 - 3/4))));
+                        auto tmp3 = static_cast<long>(4*(i2 % (ks2 / 4)));

-    buf0 = empty_strided((s1, 16*s2, (math.ceil(((-3/4) + ((1/4)*s3))))*(math.ceil(((-3/4) + ((1/4)*s3))))), (16*s2*((math.ceil(((-3/4) + ((1/4)*s3))))*(math.ceil(((-3/4) + ((1/4)*s3))))), (math.ceil(((-3/4) + ((1/4)*s3))))*(math.ceil(((-3/4) + ((1/4)*s3)))), 1), device='cpu', dtype=torch.float32)
+    buf0 = empty_strided((s1, 16*s2, (s3 // 4)*(s3 // 4)), (16*s2*((s3 // 4)*(s3 // 4)), (s3 // 4)*(s3 // 4), 1), device='cpu', dtype=torch.float32)
```

**Review by ezyang — APPROVED** [review]
I did not carefully verify the integer arithmetic'ification of the logic, but it looks plausible

**Pull request closed without merging**